### PR TITLE
Bingo Contract Assignment

### PR DESCRIPTION
**Pull request by Gagan2095**
Created a smart contract for the Bingo game. added for the necessary function as well. Also, Write an interface for the ERC-20 token which will be used in paying entry fees. A hardhat test file that contains some tests for smart contract functions.

### CODE REVIEW

File: contracts/Bingo.sol [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+    uint8 private gameIndex;
+    address private owner;
+
+    IERC20 erc20;
+
+    error InsufficientFund();
+    error Unauthorized(address);
+    error GameAlreadyStarted();
+    error ZeroUserFound();
```
Comment (schrute08) [inline contracts/Bingo.sol:41]: add natspec for errors

File: contracts/Bingo.sol [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+    /// @notice open zeppelin initializer function for transparent proxy
```
Comment (schrute08) [inline contracts/Bingo.sol:43]: add natspec for params

File: contracts/Bingo.sol [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        owner = msg.sender;
+    }
+
+    /// @notice create a new game and add in "games" mapping
+    function createGame(
+        uint8 _noOfUsers,
+        uint8 _entryFee,
+        uint256 _startDuration,
+        uint8 _turnDuration
+    ) public {
```
Comment (schrute08) [inline contracts/Bingo.sol:58]: add natspec for params

Comment (schrute08) [inline contracts/Bingo.sol:58]: make the function external

File: contracts/Bingo.sol [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        if (_noOfUsers <= 0)
```
Comment (schrute08) [inline contracts/Bingo.sol:59]: optimize the check

File: contracts/Bingo.sol [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        uint8 playersLimit;
+        uint8 generatedNumber;
+        address owner;
+        address winner;
+    }
+    mapping(uint8 gameId => GameInfo) public games;
+    mapping(uint8 gameId => mapping(address user => uint8[5][5] board)) public boards;
+    mapping(uint8 gameId => mapping(address user => uint8[5][5] board)) public drawnBoards;
+    mapping(uint8 gameId => mapping(address => bool)) public players;
+    mapping(address creator => uint8 gameId) public creatorGame;
```
Comment (schrute08) [inline contracts/Bingo.sol:31]: add natspec

File: .DS_Store
[the remark this answers is not in the document]
Comment (Gagan2095) [reply .DS_Store:1]: Done

File: .vscode/settings.json
[the remark this answers is not in the document]
Comment (Gagan2095) [reply .vscode/settings.json:1]: Done

File: .openzeppelin/sepolia.json
[the remark this answers is not in the document]
Comment (Gagan2095) [reply .openzeppelin/sepolia.json:1]: @schrute08 we should track the public network files in version control and avoid the local network files. Here the sepolia.json is a public network file hence, we should not add it to gitignore.
Here is the OpenZeppelin Docs reference
Screenshot 2024-09-17 at 8 42 04 AM

File: contracts/Bingo.sol (hunk 2 above)
Comment (Gagan2095) [reply contracts/Bingo.sol:71]: Done

File: contracts/Bingo.sol (hunk 3 above)
Comment (Gagan2095) [reply contracts/Bingo.sol]: Done

Comment (Gagan2095) [reply contracts/Bingo.sol]: Done

File: contracts/Bingo.sol (hunk 5 above)
Comment (Gagan2095) [reply contracts/Bingo.sol]: Done

File: contracts/Bingo.sol (hunk 4 above)
Comment (Gagan2095) [reply contracts/Bingo.sol]: Added modifiers for checks.

File: contracts/Bingo.sol (hunk 1 above)
Comment (Gagan2095) [reply contracts/Bingo.sol]: Done

**Pull request merged**